### PR TITLE
feat: Swagger API and semi-auto generation

### PR DESCRIPTION
**Pull request by weiseng18**
### Summary of changes

- API documentation is simpler (example below)
- Users can test endpoints at the Swagger tab of the docs
- BREAKING: All API endpoints now expect a `z.object()` at `input`
- BREAKING: tRPC endpoints are mounted at `/trpc`

### To-dos

- [x] Fix frontend API calls (all API endpoints now expect a `z.object` at `input`)
- [x] Fill up all API endpoints with the docs
- [x] Clean data at `ModuleFull.fulfillRequirements` and `ModuleFull.aliases`
- [ ] Disable create/delete interaction at Swagger
- [ ] Refactor: use `base` or not at all

### Issues

- [Generated YAML file $refs are not pointing correctly](https://github.com/jlalmes/trpc-openapi/issues/37). As a result, we can't share variables within the same `zod` object
- Doesn't seem to be possible to add entity schemas to the auto-generated OpenAPI docs, like [the ones at the bottom of this page](https://api.nusmods.com/v2/#/Modules)

### How it works

1. Describe an API endpoint in the following manner.
2. This gets converted into a OpenAPI spec yaml file. Currently, this runs every time the server refreshes.
3. The yaml file is passed into `<SwaggerUI />`
4. The component displays Swagger docs (which includes an interface for people to test routes)

```ts
  /** get a module by id */
  .query('module', {
    meta: {
      openapi: {
        enabled: true,
        tags: ['Module'],
        method: 'GET',
        path: '/module/{moduleCode}',
        summary: 'Get a single module',
      },
    },
    input: z.object({
      moduleCode: base.moduleCode,
    }),
    output: entities.Module,
    async resolve({ input }) {
      return api.moduleRepo.findByCode(input.moduleCode)
    },
  })

```

### CODE REVIEW

File: apps/server/src/app.ts [hunk 1]
```diff
@@ -2,10 +2,14 @@ import cors from 'cors'
...
+import { trpcDocumentation } from './openapi'
 
 const corsOpts = {
   origin: [
     'http://localhost:3000',
+    'http://localhost:3001',
+    'https://modtree-docs.vercel.app',
     'https://modtree-dev.vercel.app',
     'https://modtree.vercel.app',
   ],
```
Comment (weiseng18) [inline apps/server/src/app.ts:15]: allow Swagger to send API requests

File: apps/server/src/app.ts [hunk 2]
```diff
@@ -21,7 +25,12 @@ export function getApp(): Express {
   const app = express()
   app.use(cors(corsOpts))
   app.use(express.json())
-  app.use('/', createMid({ router: appRouter }))
+  app.use('/trpc', createMid({ router: appRouter }))
+  app.use('/api', createOpenApiExpressMiddleware({ router: appRouter }))
```
Comment (weiseng18) [inline apps/server/src/app.ts:29]: need separate root routes to prevent clashing

File: apps/server/src/trpc/list.ts [hunk 3]
```diff
@@ -1,39 +1,88 @@
...
+        tags: ['Module'],
+        method: 'GET',
+        path: '/modules',
+        summary: 'Get many modules',
+        description: 'Expects a comma separated string.',
+      },
+    },
+    input: z.object({
+      moduleCodes: z.string().transform(parseCommaSeparatedString),
+    }),
```
Comment (weiseng18) [inline apps/server/src/trpc/list.ts:23]: `z.preprocess()` doesn't seem to work here
- using `z.preprocess()` generates a URL like `/modules?moduleCodes=MA2001&moduleCodes=CS1010`
- only the first variable gets captured

this method
- generates a URL like `/modules?moduleCodes=MA2001,CS1010`
- first variable gets captured, but there is only one

File: apps/web/store/functions.ts [hunk 4]
```diff
@@ -190,8 +192,10 @@ export function renameGraph(graphId: string, title: string) {
  */
 export function setBuildTarget(degreeId: string) {
   trpc
-    .query('degree', degreeId)
-    .then((degree) => trpc.query('modules', degree.modules))
+    .query('degree', { degreeId })
+    .then((degree) =>
+      trpc.query('modules', { moduleCodes: degree.modules.join(',') })
+    )
```
Comment (weiseng18) [inline apps/web/store/functions.ts:198]: downside of parsing in the `resolve()`: frontend must pass in a comma separated string, rather than an array

**Comment by jlalmes** [comment]
Hi @weiseng18 - I've just opened an issue for the `trpc-openapi@v1.0.0` roadmap (https://github.com/jlalmes/trpc-openapi/issues/91), I believe it addresses many of the issues you've raised here.

> https://github.com/jlalmes/trpc-openapi/issues/37. As a result, we can't share variables within the same zod object.

Could you help me understand what you mean by `sharing variables within the same zod object`?

**Comment by weiseng18** [comment]
split up this PR as per #378

**Pull request closed without merging**